### PR TITLE
feat: add negative topups

### LINKED ISSUE #2711: [Feature request] Deduct wallet button, opposite of topup.
Sometimes I want to lower the balance of a wallet. Current workaround is to send sats to a different wallet, shifting the problem there.

As admin/superuser I would like to have a button that does the opposite of Topup; deduct x sats from the wallet.

### PR DESCRIPTION
**Pull request by dni**
closes #2711

### CODE REVIEW

File: lnbits/core/services/payments.py [hunk 1]
```diff
@@ -195,21 +196,54 @@ def service_fee(amount_msat: int, internal: bool = False) -> int:
...
+    if amount == 0:
+        raise ValueError("Amount cannot be 0.")
+    if (
+        settings.lnbits_wallet_limit_max_balance > 0
+        and wallet.balance + amount > settings.lnbits_wallet_limit_max_balance
+    ):
+        raise ValueError("Balance change failed, amount exceeds maximum balance.")
+
+    # negative topup
+    if amount < 0:
```
Comment (motorina0) [inline lnbits/core/services/payments.py:214]: - this `if` can be omitted

File: lnbits/core/services/payments.py [hunk 2]
```diff
@@ -195,21 +196,54 @@ def service_fee(amount_msat: int, internal: bool = False) -> int:
...
+                wallet_id=wallet.id,
+                amount=amount,
+                memo="Admin top up" if amount < 0 else "Admin withdrawal",
+                internal=True,
+                conn=conn,
+            )
+            payment.status = PaymentState.SUCCESS
+            await update_payment(payment, conn=conn)
+            # notify receiver asynchronously
+            from lnbits.tasks import internal_invoice_queue
```
Comment (motorina0) [inline lnbits/core/services/payments.py:230]: - we should avoid import in the middle of the code, can the queue be an optional input param instead?

File: lnbits/core/views/user_api.py [hunk 3]
```diff
@@ -269,14 +269,21 @@ async def api_users_delete_user_wallet(user_id: str, wallet: str) -> SimpleStatu
 @users_router.put(
     "/topup",
     name="Topup",
+    summary="Increase balance for a particular wallet.",
+    dependencies=[Depends(check_super_user)],
+    deprecated=True,
```
Comment (motorina0) [inline lnbits/core/views/user_api.py:274]: - lets just remove this path for `v1`

File: lnbits/core/views/user_api.py [hunk 4]
```diff
@@ -269,14 +269,21 @@ async def api_users_delete_user_wallet(user_id: str, wallet: str) -> SimpleStatu
...
-    await get_wallet(data.id)
-    if settings.lnbits_backend_wallet_class == "VoidWallet":
-        raise Exception("VoidWallet active")
-
-    await update_wallet_balance(wallet_id=data.id, amount=int(data.amount))
+async def api_update_balance(data: CreateTopup) -> SimpleStatus:
+    wallet = await get_wallet(data.id)
+    if not wallet:
+        raise HTTPException(
+            status_code=HTTPStatus.NOT_FOUND, detail="Wallet not found."
```
Comment (motorina0) [inline lnbits/core/views/user_api.py:286]: nitpick: constructor keys (`status_code`, `status_code`) can be omitted, it makes the code a little less verbouse

File: lnbits/core/services/payments.py [hunk 5]
```diff
@@ -195,21 +196,54 @@ def service_fee(amount_msat: int, internal: bool = False) -> int:
...
+    if amount < 0:
+        if wallet.balance_msat + amount < 0:
+            raise ValueError("Balance change failed, can not go into negative balance.")
+
+    async with db.reuse_conn(conn) if conn else db.connect() as conn:
+        if amount > 0:
+            payment = await create_invoice(
+                wallet_id=wallet.id,
+                amount=amount,
+                memo="Admin top up" if amount < 0 else "Admin withdrawal",
```
Comment (motorina0) [inline lnbits/core/services/payments.py:223]: redundant: we know `amount > 0`

File: lnbits/core/services/payments.py [hunk 6]
```diff
@@ -195,21 +196,54 @@ def service_fee(amount_msat: int, internal: bool = False) -> int:
...
+            from lnbits.tasks import internal_invoice_queue
+
+            await internal_invoice_queue.put(payment.checking_id)
+        else:
+            payment_hash = random_hash()
+            await create_payment(
+                checking_id=f"internal_{payment_hash}",
+                data=CreatePayment(
+                    wallet_id=wallet.id,
+                    bolt11="fake_bolt11",
```
Comment (motorina0) [inline lnbits/core/services/payments.py:239]: it should be a valid bolt with a `"fake_bolt11"` memo

Comment (motorina0) [reply lnbits/core/services/payments.py:239]: The invoice should not be created by the funding source, but just generated with the `bolt11` lib.

I imagine a `sanity check` script that goes over all the payments and checks the `bolt11` against the actual payment `amount` and `memo`. In this case the valid `bolt11` would help.

File: lnbits/core/services/payments.py (hunk 2 above)
Comment (dni) [reply lnbits/core/services/payments.py]: can you look at that in another PR? it was like that before and its related to circular imports

Comment (motorina0) [reply lnbits/core/services/payments.py]: ok

**Review by talvasconcelos — APPROVED** [review]
It works as expected, tested!

Only thing is that it's not clear how to "reverse" top up! It's not clear that you need to enter a negative amount

**Review by motorina0 — APPROVED** [review]

**Pull request merged**